app/jobs/update_from_saml_service.rb

```diff
@@ -0,0 +1,161 @@
+# frozen_string_literal: true
+
+require 'net/http'
+
+class UpdateFromSAMLService
+  include GetSAMLMetadata
+
+  def self.perform
+    new.perform
+  end
+
+  def perform
+    @service_providers = []
+    @identity_providers = []
+
+    doc = document(source, metadata_cert)
+    ActiveRecord::Base.transaction do
+      doc.xpath('//md:EntityDescriptor', SAML_NAMESPACES).each do |node|
+        process_entity(node)
+      end
+
+      clean
+    end
+
+    nil
+  end
+
+  def process_entity(node)
+    entity_id = node.attributes['entityID'].value
+
+    org_node = xpath_at(node, './md:Organization')
+    org = process_org(org_node) if org_node
+    return unless org
+
+    reg_info_node = xpath_at(node, './md:Extensions/mdrpi:RegistrationInfo')
+    reg_date = attr_val(reg_info_node, 'registrationInstant') if reg_info_node
+
+    idp_node = xpath_at(node, './md:IDPSSODescriptor')
+    process_idp(idp_node, entity_id, org, reg_date) if idp_node
+
+    sp_node = xpath_at(node, './md:SPSSODescriptor')
+    process_sp(sp_node, entity_id, org, reg_date) if sp_node
+  end
+
+  def process_org(node)
+    org_id_node = xpath_at(node, './md:OrganizationName')
+    org_name_node = xpath_at(node, './md:OrganizationDisplayName')
+
+    return nil unless org_id_node && org_name_node
+
+    org_id = org_id_node.content.strip
+    org_name = org_name_node.content.strip
+
+    org = Organization.find_or_initialize_by(domain: org_id)
+    org_attrs = { name: org_name }
+
+    org_attrs[:identifier] = org_identifier(org_id) if org.id.nil?
+
+    org.update!(org_attrs)
+
+    org
+  end
+
+  def process_idp(node, entity_id, org, reg_date)
+    idp = IdentityProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
```

Comment:
Watch out - there might be multiple instances of `mdui:DisplayName`, distinguished by the `xml:lang` attribute.  You'd want `xml:lang="en"`.  In Tuakiri, you'd likely only see the `en` versions, but eduGAIN-imported metadata might have more languages.

Reply:
Sorted.